linuxkm/lkcapi_glue.c

```diff
@@ -723,14 +725,34 @@ static int km_AesGcmDecrypt(struct aead_request *req)
     }
 
     err = wc_AesGcmInit(ctx->aes_encrypt, NULL /*key*/, 0 /*keylen*/, walk.iv,
-                        WC_AES_BLOCK_SIZE);
+                        GCM_NONCE_MID_SZ);
     if (unlikely(err)) {
         pr_err("%s: wc_AesGcmInit failed: %d\n",
                crypto_tfm_alg_driver_name(crypto_aead_tfm(tfm)), err);
         return -EINVAL;
     }
 
-    assoc = scatterwalk_map(&assocSgWalk);
+    if (req->src->length >= req->assoclen && req->src->length) {
+        scatterwalk_start(&assocSgWalk, req->src);
+        assoc = scatterwalk_map(&assocSgWalk);
+        if (unlikely(IS_ERR(assoc))) {
+            pr_err("%s: scatterwalk_map failed: %ld\n",
+                   crypto_tfm_alg_driver_name(crypto_aead_tfm(tfm)),
+                   PTR_ERR(assoc));
+            return err;
+        }
+    }
+    else {
+        /* assoc can be any length, so if it's noncontiguous, we have to copy it
+         * to a contiguous heap allocation.
+         */
+        assocmem = malloc(req->assoclen);
+        if (unlikely(assocmem == NULL))
+            return -ENOMEM;
+        assoc = assocmem;
+        scatterwalk_map_and_copy(assoc, req->src, 0, req->assoclen, 0);
+    }
+
     if (unlikely(IS_ERR(assoc))) {
         pr_err("%s: scatterwalk_map failed: %ld\n",
                crypto_tfm_alg_driver_name(crypto_aead_tfm(tfm)),
```

Comment:
If we return err here, we leak assocmem if it had been malloced?

Reply:
Also, I think it's impossible for `IS_ERR(assoc)` to be true here?

Reply:
fixed